File: backend/src/shared/utils/error.util.ts
import {
	BadRequestException,
	ConflictException,
	ForbiddenException,
	InternalServerErrorException,
	NotAcceptableException,
	NotFoundException,
	UnauthorizedException,
} from '@nestjs/common';

import { ErrorType } from '../types/error.type';
import { ErrorResponseType } from '../types/return.type';
import { isObject } from './object.util';

export function throwError(props: { error: string; errorType: ErrorType }): never {
	const { error, errorType } = props;

	const errorObject: { [key in ErrorType]: any } = {
		BadRequestException: new BadRequestException(error),
		ConflictException: new ConflictException(error),
		ForbiddenException: new ForbiddenException(error),
		InternalServerErrorException: new InternalServerErrorException(error),
		NotAcceptableException: new NotAcceptableException(error),
		NotFoundException: new NotFoundException(error),
		UnauthorizedException: new UnauthorizedException(error),
	};

	throw errorObject[errorType] || new InternalServerErrorException('An unexpected error occurred');
}

export const isError = (response: unknown): response is ErrorResponseType => {
	return isObject(response) && 'error' in response;
};

File: ai-server/app/shared/utils/error_util.py
from fastapi import HTTPException, status
from pydantic import BaseModel
from typing import Any, TypeGuard

from app.shared.types.return_type import ErrorResponseType
from app.shared.types.error_type import ErrorType


class ThrowErrorArgs(BaseModel):
    error: str
    errorType: ErrorType


def throw_error(args: ThrowErrorArgs):
    if (args.errorType == "BadRequestException"):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=args.error
        )
    elif (args.errorType == "ConflictException"):
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=args.error
        )
    elif (args.errorType == "ForbiddenException"):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=args.error
        )
    elif (args.errorType == "InternalServerErrorException"):
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=args.error
        )
    elif (args.errorType == "NotAcceptableException"):
        raise HTTPException(
            status_code=status.HTTP_406_NOT_ACCEPTABLE,
            detail=args.error
        )
    elif (args.errorType == "NotFoundException"):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=args.error
        )
    elif (args.errorType == "UnauthorizedException"):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail=args.error
        )
    else:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=args.error
        )


def is_error(response: Any) -> TypeGuard[ErrorResponseType]:
    return isinstance(response, ErrorResponseType)


def carry_error(error: ErrorResponseType, trace: str) -> ErrorResponseType:
    return ErrorResponseType(
        **error.model_dump(exclude={"trace"}),
        trace=error.trace + [trace]
    )
